IdentityCore/src/requests/sdk/msal/MSIDDefaultBrokerResponseHandler.m

```diff
@@ -22,25 +22,184 @@
+    {
+        return [[MSIDAADV2BrokerResponse alloc] initWithDictionary:decryptedResponse error:error];
+    }
+    
+    // Failure case
+    MSIDAADV2BrokerResponse *brokerResponse = [[MSIDAADV2BrokerResponse alloc] initWithDictionary:decryptedResponse error:error];
+    
+    if (!brokerResponse)
+    {
+        return nil;
+    }
+    
+    NSError *brokerError = [self resultFromBrokerErrorResponse:brokerResponse
+                                             userDisplayableId:userDisplayableId];
+    
+    if (error)
+    {
+        *error = brokerError;
+    }
+    
     return nil;
 }
 
 - (id<MSIDCacheAccessor>)cacheAccessorWithKeychainGroup:(NSString *)keychainGroup
                                                   error:(NSError **)error
 {
-    // TODO: MSAL pieces
+#if TARGET_OS_IPHONE
+    MSIDKeychainTokenCache *dataSource = [[MSIDKeychainTokenCache alloc] initWithGroup:keychainGroup];
+    MSIDLegacyTokenCacheAccessor *otherAccessor = [[MSIDLegacyTokenCacheAccessor alloc] initWithDataSource:dataSource otherCacheAccessors:nil factory:self.oauthFactory];
+    MSIDDefaultTokenCacheAccessor *cache = [[MSIDDefaultTokenCacheAccessor alloc] initWithDataSource:dataSource otherCacheAccessors:@[otherAccessor] factory:self.oauthFactory];
+    return cache;
+#else
+    if (error)
+    {
+        *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Broker responses not supported on macOS", nil, nil, nil, nil, nil);
+    }
+    
     return nil;
+#endif
 }
 
+- (NSError *)resultFromBrokerErrorResponse:(MSIDAADV2BrokerResponse *)errorResponse
+                         userDisplayableId:(NSString *)userId
+{
+    NSString *errorDomain = errorResponse.errorDomain ?: MSIDErrorDomain;
+    
+    NSString *errorCodeString = errorResponse.errorCode;
+    NSInteger errorCode = MSIDErrorBrokerUnknown;
+    if (errorCodeString && ![errorCodeString isEqualToString:@"0"])
+    {
+        errorCode = [errorCodeString integerValue];
+    }
+    
+    NSString *errorDescription = errorResponse.errorDescription;
+    if (!errorDescription)
+    {
+        errorDescription = @"Broker did not provide any details";
+    }
+    
+    NSString *oauthErrorCode = errorResponse.oauthErrorCode;
+    NSString *subError = errorResponse.subError;
+    NSUUID *correlationId = [[NSUUID alloc] initWithUUIDString:errorResponse.correlationId];
+    
+    NSMutableDictionary *userInfo = [NSMutableDictionary new];
+    for (NSString * metadataKey in errorResponse.errorMetadata.allKeys)
+    {
+        NSString *userInfokey = _userInfoKeyMapping[metadataKey];
+        if (userInfokey)
+        {
+            [userInfo setValue:errorResponse.errorMetadata[metadataKey] forKey:userInfokey];
+        }
+    }
+
+    userInfo[MSIDUserDisplayableIdkey] = errorResponse.userId ? errorResponse.userId : userId;
+
+    NSDictionary *httpHeaders = [NSDictionary msidDictionaryFromWWWFormURLEncodedString:errorResponse.httpHeaders];
```

Comment:
Wouldn't http headers be part of error metadata?

Reply:
The for-loop at line 181-187 can only handle string-type metadata, which is the case for most properties. Http headers are dictionary type, so need some special handling after the for-loop.

Added some comments to clarify the above :)

Reply:
Ok, but shouldn't it come from errorResponse.errorMetadata?

Reply:
Yep, the getter retrieves the value from `errorResponse.errorMetadata` :)

Reply:
👍